### PR TITLE
implementation of adding load by geometry id with tests.

### PR DESCRIPTION
**Pull request by morettid**

### CODE REVIEW

File: tests/test_model.py [hunk 1]
```diff
@@ -977,45 +977,6 @@ def test_add_line_load_to_3_edges(self, expected_geometry_line_load: Geometry,
 
         TestUtils.assert_almost_equal_geometries(expected_geometry, generated_geometry)
 
-    def test_add_moving_point_load(self, expected_geometry_line_load: Geometry,
```
Comment (aronnoordam) [inline tests/test_model.py:980]: why did you remove this test? It should still work

File: tests/test_model.py [hunk 2]
```diff
@@ -1501,6 +1461,96 @@ def test_add_boundary_condition_by_geometry_ids(self, create_default_3d_soil_mat
...
+                                      create_default_line_load_parameters: LineLoad,
+                                      create_default_surface_load_parameters: SurfaceLoad):
+        """
+        Test if a load is added correctly to the model. Here the load is added to the model by
+        specifying the geometry ids to which the load should be applied.
+
+        Args:
+            - create_default_3d_soil_material (:class:`stem.soil_material.SoilMaterial`): A default soil material.
+            - create_default_point_load_parameters (:class:`stem.load.PointLoad`): default point load parameters.
+            - create_default_moving_load_parameters (:class:`stem.load.MovingLoad`): default moving load parameters.
```
Comment (aronnoordam) [inline tests/test_model.py:1475]: missing surface load in docstring. Also create_default_moving_load_parameters  is not present in the input

File: tests/test_model.py [hunk 3]
```diff
@@ -1501,6 +1461,96 @@ def test_add_boundary_condition_by_geometry_ids(self, create_default_3d_soil_mat
...
+        npt.assert_allclose(model.process_model_parts[1].parameters.value, expected_1d_model_part_parameters.value)
+        npt.assert_allclose(model.process_model_parts[1].parameters.active, expected_1d_model_part_parameters.active)
+
+        # check 2d parameters
+        npt.assert_allclose(model.process_model_parts[2].parameters.value, expected_2d_model_part_parameters.value)
+        npt.assert_allclose(model.process_model_parts[2].parameters.active, expected_2d_model_part_parameters.active)
+
+        # check raising of errors
+
+        with pytest.raises(ValueError):
```
Comment (aronnoordam) [inline tests/test_model.py:1544]: please also check the value error messages

Comment (aronnoordam) [reply tests/test_model.py:1544]: also please create a new test for this. In order to keep the tests smaller

File: stem/model.py [hunk 4]
```diff
@@ -149,6 +149,49 @@ def add_soil_layer_by_coordinates(self, coordinates: Sequence[Sequence[float]],
...
+        # add physical group to gmsh
+        self.gmsh_io.add_physical_group(name, ndim_load, geometry_ids)
+
+        # create model part
+        model_part = ModelPart(name)
+
+        # retrieve geometry from gmsh and add to model part
+        model_part.get_geometry_from_geo_data(self.gmsh_io.geo_data, name)
+
+        # add boundary parameters to model part
```
Comment (aronnoordam) [inline stem/model.py:190]: wrong comment

**Review by aronnoordam — CHANGES_REQUESTED** [review]

File: tests/test_model.py (hunk 1 above)
Comment (morettid) [reply tests/test_model.py]: Honestly, I overlooked it!

File: tests/test_model.py [hunk 5]
```diff
@@ -1539,16 +1584,53 @@ def test_add_load_by_geometry_ids(self, create_default_3d_soil_material: SoilMat
         npt.assert_allclose(model.process_model_parts[2].parameters.value, expected_2d_model_part_parameters.value)
         npt.assert_allclose(model.process_model_parts[2].parameters.active, expected_2d_model_part_parameters.active)
 
-        # check raising of errors
+        # add boundary conditions in 0d, 1d and 2d
+        model.add_load_by_geometry_ids([], create_default_point_load_parameters, "point_loads")
```
Comment (aronnoordam) [inline tests/test_model.py:1588]: you don't assert anything after this adding of load, also the comment is not correct.

also moving load is not tested

File: requirements.txt [hunk 6]
```diff
@@ -1,3 +1,4 @@
 numpy==1.24.2
 plotly==5.15.0
+networkx==3.1
```
Comment (aronnoordam) [inline requirements.txt:3]: i dont think this package is needed anymore, see comments below

File: stem/utils.py [hunk 7]
```diff
@@ -371,5 +373,99 @@ def create_sigmoid_tiny_expr(start_time: float, dt_slope: float, initial_value:
...
+                raise ValueError(f"Found {num_loops} loop(s) in the path.")
+
+            if num_bp > 0:
+                raise ValueError(f"Path is branching, should be on a line."
+                                 f"{num_bp} branching point(s) have been found in the path!")
+
+    @staticmethod
+    def is_point_aligned_and_between_any_of_points(coordinates: Sequence[Sequence[Sequence[float]]],
+                                                   origin: Sequence[float]):
+        """Checks that any of the points provides in a list of pairs of coordinates
```
Comment (aronnoordam) [inline stem/utils.py:443]:
```suggestion
        """
        Checks that any of the points provides in a list of pairs of coordinates
```

File: stem/utils.py [hunk 8]
```diff
@@ -371,5 +373,99 @@ def create_sigmoid_tiny_expr(start_time: float, dt_slope: float, initial_value:
...
+            - coordinates (Sequence[Sequence[Sequence[float]]]): Pair-wise sets of coordinates representing the line
+                on which the origin should lie.
+            - origin (Sequence[float]): the coordinates of the point to be checked for alignment.
+
+        Raises:
+            - ValueError: when point is not aligned with one of the lines (pair-wise sets of coordinates).
+
+        """
+
+        for ix in range(len(coordinates)):
```
Comment (aronnoordam) [inline stem/utils.py:455]:
```suggestion
        for line in coordinates:
```

File: stem/utils.py [hunk 9]
```diff
@@ -371,5 +373,99 @@ def create_sigmoid_tiny_expr(start_time: float, dt_slope: float, initial_value:
...
+            )
+            # check origin is between the edges of the line (edges included)
+            is_between_check = Utils.is_point_between_points(
+                point=origin, start_point=coordinates[ix][0], end_point=coordinates[ix][1]
+            )
+            # check if point complies
+            is_on_line = collinear_check and is_between_check
+            # exit at the first success of the test (point in the line)
+            if is_on_line:
+                return
```
Comment (aronnoordam) [inline stem/utils.py:468]: The name of the function suggests that a bool is returned. I would split the check and the validation

File: tests/test_model.py [hunk 10]
```diff
@@ -1501,6 +1498,141 @@ def test_add_boundary_condition_by_geometry_ids(self, create_default_3d_soil_mat
...
+            4: Surface.create([8, 11, -10, -3], 4)
+        }
+        # collect all expected geometries
+        all_expected_geometries = [expected_load_geometry_0d, expected_load_geometry_1d, expected_load_geometry_2d]
+
+        for expected_geometry, model_part in zip(all_expected_geometries, model.process_model_parts):
+
+            TestUtils.assert_almost_equal_geometries(expected_geometry, model_part.geometry)
+
+        # check 0d parameters
```
Comment (aronnoordam) [inline tests/test_model.py:1575]: I would change these comments and parameter names refering to the load type rather than the geometry dimension. In this case " Check point load parameters " and  npt.assert_allclose(model.process_model_parts[0].parameters.value, expected_point_load_parameters.value)

File: tests/test_utils.py [hunk 11]
```diff
@@ -458,3 +459,119 @@ def test_create_sigmoid_tiny_expr(self):
...
+        }
+
+        # test for branching points
+        msg = "Path is branching, should be on a line.1 branching point\(s\) have been found in the path!"
+        with pytest.raises(ValueError, match=msg):
+            Utils.check_lines_geometry_are_path(geo4)
+
+    def test_is_point_aligned_and_between_any_of_points(self):
+        """Checks that any of the points is aligned with at least one of the points in a list of pairs of
+         coordinates"""
```
Comment (aronnoordam) [inline tests/test_utils.py:562]:
```suggestion
        """
        Checks that any of the points is aligned with at least one of the points in a list of pairs of
         coordinates
         """
```

File: stem/model.py [hunk 12]
```diff
@@ -149,6 +148,71 @@ def add_soil_layer_by_coordinates(self, coordinates: Sequence[Sequence[float]],
...
+                line_coords = []
+                for k in line.point_ids:
+                    line_coords.append(model_part.geometry.points[k].coordinates)
+                coordinates.append(line_coords)
+
+            # check origin of moving load is in the path
+            Utils.is_point_aligned_and_between_any_of_points(coordinates, load_parameters.origin)
+            # check that the path provided by geometry is correct (no loops, no branching out
+            # and no discontinuities in the path)
+            Utils.check_lines_geometry_are_path(model_part.geometry)
```
Comment (aronnoordam) [inline stem/model.py:209]: i would return a bool from this functions, and put the validation here

**Review by aronnoordam — CHANGES_REQUESTED** [review]

File: stem/utils.py [hunk 13]
```diff
@@ -371,5 +372,103 @@ def create_sigmoid_tiny_expr(start_time: float, dt_slope: float, initial_value:
...
+                if len(line_to_point) > 2:
+                    # "Branching point was found for node {p} connected to lines {line_to_point}
+                    return False
+
+            # if no branching point are found than the check of connectivity holds when
+            if len(unique_points) != (len(lines) + 1):
+                # lines are not connected.
+                return False
+
+        # if no False is return, the path is valid.  Return True.
```
Comment (aronnoordam) [inline stem/utils.py:437]:
```suggestion
        # All lines are connected without branches
```

File: setup.cfg [hunk 14]
```diff
@@ -19,6 +19,7 @@ install_requires =
     numpy>=1.24
     matplotlib>=3.7.1
     plotly>=5.15.0
+    networkx>=3.1
```
Comment (aronnoordam) [inline setup.cfg:22]:
```suggestion
```

**Review by aronnoordam — APPROVED** [review]
